graphql_compiler/cost_estimation/filter_selectivity_utils.py

```diff
@@ -44,6 +60,359 @@ def _are_filter_fields_uniquely_indexed(filter_fields, unique_indexes):
+        stronger_lower_bound = lower_bound_a
+    elif lower_bound_b is not None:
+        # If only one lower bound exists, the stronger lower bound equals it.
+        stronger_lower_bound = lower_bound_b
+
+    return stronger_lower_bound
+
+
+def _get_stronger_upper_bound(upper_bound_a, upper_bound_b):
+    """Return the smaller bound of the two given upper bounds.
+
+    Args:
+        upper_bound_a: int or None, describing one of the upper bounds. If the bound does not exist,
+                       the argument value should be None.
+        upper_bound_b: int or None, describing one of the upper bounds. If the bound does not exist,
+                       the argument value should be None.
+
+    Returns:
+        - int, the smaller of the two upper bounds, if one or more upper bounds have an integer
+          value.
+        - None if both upper bounds have a value of None.
+    """
+
+    stronger_upper_bound = None
+    if upper_bound_a is not None and upper_bound_b is not None:
+        # If both upper bounds exist, the stronger upper bound is the smaller bound of the two
+        # given upper bounds.
+        stronger_upper_bound = min(upper_bound_a, upper_bound_b)
+    elif upper_bound_a is not None:
+        # If only one upper bound exists, the stronger upper bound equals it.
+        stronger_upper_bound = upper_bound_a
+    elif upper_bound_b is not None:
+        # If only one upper bound exists, the stronger upper bound equals it.
+        stronger_upper_bound = upper_bound_b
+
+    return stronger_upper_bound
+
+
+def _get_intersection_of_intervals(interval_a, interval_b):
+    """Return the intersection of two IntegerIntervals, or None if the intervals are disjoint.
+
+    Args:
+        interval_a: IntegerInterval namedtuple.
+        interval_b: IntegerInterval namedtuple.
+
+    Returns:
+        - IntegerInterval namedtuple, intersection of the two given IntegerIntervals, if the
+          intersection is not empty.
+        - None otherwise.
+    """
+    strong_lower_bound = _get_stronger_lower_bound(interval_a.lower_bound, interval_b.lower_bound)
+    strong_upper_bound = _get_stronger_upper_bound(interval_a.upper_bound, interval_b.upper_bound)
+
+    intersection = _create_integer_interval(strong_lower_bound, strong_upper_bound)
+    return intersection
+
+
+def _get_query_interval_of_binary_integer_inequality_filter(
+    parameter_values, filter_operator
+):
+    """Return IntegerInterval or None of values passing through a binary integer inequality filter.
+
+    Args:
+        parameter_values: List[int], describing the parameters for the inequality filter.
+        filter_operator: str, describing the binary inequality filter operation being performed.
+
+    Returns:
+        - IntegerInterval namedtuple, non-empty interval of values that pass through the filter.
+        - None if the interval is empty.
+
+    Raises:
+        - AssertionError if the given operator is unsupported.
```

Comment:
I'm not sure if others would agree, but I think documenting assertion errors doesn't make much sense:
1. An assertion error is raised when the preconditions aren't met. You're not required to document undefined behavior.
2. Nobody needs this documentation. Nobody should be catching assertion errors.
3. If I had to document every assertion error I raise, I might be discouraged to make assertions. Python code I've seen most often has fewer assertions that I'd like it to.